uint32 vector_len
point[] vector_point